File: medium/cpp/1973.cpp
// https://leetcode.com/problems/count-nodes-equal-to-sum-of-descendants/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int equalToDescendants(TreeNode* root)
    {
        auto res = 0;
        static_cast<void>(count(root, res));
        return res;
    }

private:
    long long count(TreeNode* root, int& res)
    {
        if (root == nullptr)
        {
            return 0;
        }

        const auto sum = 0LL + count(root->left, res) + count(root->right, res);
        res += sum == root->val;
        return sum + root->val;
    }
};

File: medium/python/1973.py
# https://leetcode.com/problems/count-nodes-equal-to-sum-of-descendants/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def equalToDescendants(self, root: Optional[TreeNode]) -> int:
        def solve(root):
            if root is None:
                return 0, 0
            left = solve(root.left)
            right = solve(root.right)
            tot = left[1] + right[1]

            if root.val == left[0] + right[0]:
                tot += 1

            return root.val + left[0] + right[0], tot

        return solve(root)[1]        